### DISCUSSION TITLE
Don't require DNS to securely run fedimint server

### DISCUSSION THREAD

**Discussion by dpc** [fedimint/fedimint#3304] [General]
> Currently the most difficult / time-consuming thing about setting up a fedimint guardian server is probably buying a domain name, updating DNS records to point it at fedimint server, and waiting for DNS to propagate.

### Context

It's trivial to get a public IP or at least publicly forwarded port. The problem is, typically nothing will guarantee that this IP doesn't need to change with time. Servers need to be migrated etc.

So one can think about DNS as being a mapping from an immutable "ID" of some kind to an IP. And importantly **mutable** mapping. Everything else on the internet rests on top of this mutable database.

Unfortunately DNS is not self-publishable, and that's our biggest problem with it. We wish that `fedimintd` could produce *some* ID that can have  mutable over time mapping to IP and be self-publishable and censorship resistant. We don't really care about nicely looking human readable domains. As long as we could get some `jh9ewj98iu0jhggfsahusajh0sfda` that can be updated over time to an IP, and looked up, it would work for us.

The single biggest problem with with replacing DNS with anything is that TLS and all things HTTPS assume DNS. We can either somehow inject things into DNS namespace in some backdoor way, or just drop public DNS CA based TLS.

Now, injecting things into DNS is going to be against lots of software. It can be done by configuring system/software to point at some custom DNS resolver that does "things" and responds. Another one is using DoH (DNS over HTTP) which is sometimes configurable). Unfortunately this might not always be possible, and SSL cert authorities (like letsencrypt) won't be compatible. Which means needing to inject some custom self-CA-certs, etc.

Alternatively one can just drop trying to support general purpose software. . Only other `fedimintd`s and fedimint clients need to be able to connect to `fedimintd` servers. Especially if we include Tor connectivity for clients, then p2p communication is always `fedimintd` to `fedimintd` already anyway.

Then we can use any DNS-like system, lookup the IP to connect to and and deal with encryption ourselves. For software that can't deal with such a custom connection, we can even open a local tcp port doing the tunneling and point it at it. This way any software should be able to work.

Worth mentioning that we could have Federation Peers be able to use optionally both DNS and non-DNS connectivity.

### Solutions

Tor. Tor has everything built-in and takes care of the DNS completely. Lots of software has a support for it built in. The problem - it often doesn't work well.

[Veilid](https://veilid.com/). AFAIU Veilid is like Tor. You build it into your app, and you get Tor-like networking, that is supposed to work better in practice. Still immature, untried though. Could work both p2p and api if delivers.

Nostr or nostr-like solution. Just publish something redundantly and let other people look it up. Everything is self signed anyway. Each peer in Federation can act as a relay for other peers (and even other federations), so one can discover IPs to use for other peers as long as they can connect to at least on honest peer.

DHT. DHTs are kind of like p2p nostr. There's [Pkarr](https://github.com/Nuhvi/pkarr) which works exactly on the thing - a DNS-like system that publishes self-signed records to Bittorrent's DHT (aka Mainnet DHT). It includes DoH gateway, and possibly a DNS resolver. Right now it only has JS implementation, but author wants to rewrite in Rust. We could run it locally next to fedimintd, if it was to be used only for p2p traffic.

**Comment by dpc** [comment]
https://docs.rs/pkarr/latest/pkarr/

**Comment by Nuhvi** [comment]
Both Nostr and Veilid will force you to use specific transport, in the case of the former, their channels, and in the case of the former WebScoket. Whereas Pkarr is purely a permissionless DNS root server, once you get your A, AAA, CNAME, or TXT records, you are free to use whatever transport you want. 

Also, Nostr has no solution for routing, if I open a Relay and post to my own relay, no one can find it unless I also post to popular relays that others are watching. whereas Mainline is all connected, the best effort in Nostr is the gossip protocol, and even the author of that said that it doesn't work for generic cases, and that the ultimate solution is a DHT.

soon, I will demo using plain `dig` cli to fetch Pkarr records by just providing a relay server, but unlike in Nostr, that relay is just a transparent relay over Mainline, meaning even a relay in localhost can still find any record on the DHT.

**Comment by BitcoinErrorLog** [comment]
I am necroing this thread in light of the recent DNS mint issues occurring. Use pkarr, and consider the rest of the designs mentioned in https://github.com/synonymdev/atomicity/ -- we are trying to keep things practical

**Reply by dpc** [reply]
What do you use for web connectivity?

We generally agree that we want to drop DNS, but we need ability for web clients to interact with Federations. Broadly Pkarr + libnoise e2e encryption with tunneling over http/https is what we currently consider. But all of this will be quite a bit of engineering effort.

**Reply by Nuhvi** [reply]
For the web we are using the Relay spec https://relay.pkarr.org/, it is implemented in Rust already with cache-control headers making it very fast in browser, and sane defaults for rate limiting etc.

Sure, it is a lot of work, but only because ICANN DNS comes for free with the operating system, not because Pkarr is particularly more complex than it needs to.

So the options are either ICANN or Pkarr or something more complex than Pkarr. If I may say so myself, there is no more room for simplification.

That being said, that work doesn't need to be done 10 times, only once, and every server and every HTTP client can use it going forward.

As for HTTPs, sure self-issued certs or Noise would be great, but won't work for browsers, still would be great for mobile, but merely pointing Pkarr to CNAME is enough to get the censorship resistance; the moment your DNS rug pull you, you use another, or even have multiple CNAMEs.

Finally we will also be focusing on the key management UX of the whole thing, so hopefully using CNAME we can have key delegation so you don't lose your TLD if your server is compromised.

Yes, lots of work, but ... that is always the case, we are lucky there is a working solution at all.

**Reply by dpc** [reply]
Just to clarify - There's nothing that Pkarr does wrong, or even could do better, AFAICT. But the reality is that we're boxed with solutions only to things that work seamlessly with the web, even if just to be an escape hatch in case Fedimint apps start getting banned in the mobile stores.

Web relays for Pkarr from web client are fine. Libnoise can be run in the web client to tunnel everything trustlessly over https tunnel. That's also fine. So it is possible and doable, and we'll get there.

**Reply by Nuhvi** [reply]
Curious about your plans for Libnoise... I am also aware of some attempts to use noise with QUIC https://eprint.iacr.org/2019/028.pdf which should work for HTTP/3.

But the safer easier path for now is just to point Pkarr to a normal CNAME and use its TLS cert ... no need to fix everything at once.

**Reply by dpc** [reply]
@Nuhvi So use Pkarr to publish CNAMES resolving to normal DNS names, so that these can be changed? 

That might actually be exactly what we need. At least for time being.

How would we distinguish "normal DNS" from "DNS that needs to be resolved via Pkarr first". Use `https+pkarr://o4dksfbqk85ogzdb5osziw6befigbuxmuxkuxq8434q89uj56uyy` maybe ...

So the old clients and Federations can work with plain DNS names, while new versions can take of automatically publishing and commitint to Pkarr name in the Federation ID/config.


> Curious about your plans for Libnoise... 

If we were to dump DNS completely, we just want Pkarr name to resolve to IPs and then... we need some ability to make an authenticated connection between a web (or any other) client and server. Since we're out of DNS names, we can't do TLS, so the only option seems like tunneling something like libnoise inside a untrusted web proxy, possibly provided by the same server that hosts the web app.


I've been looking at Iroh and it looks great and could do this and many other things for us, but we would need wasm story sorted out: https://github.com/quinn-rs/quinn/issues/1388 and some level of commitment to protocol stability, which I'm not sure if is available yet.


@joschisan @elsirion Ping for awareness.

**Reply by Nuhvi** [reply]
> How would we distinguish "normal DNS" from "DNS that needs to be resolved via Pkarr first". 

The way we do it, is to test that it is 52 character first, if it is, try to convert it into a PublicKey. What are the chances that ICANN registered a TLD that is 52 character AND is a valid z-base32 encoded public key?

> If we were to dump DNS completely, we just want Pkarr name to resolve to IPs and then... we need some ability to make an authenticated connection between a web (or any other) client and server. Since we're out of DNS names, we can't do TLS, so the only option seems like tunneling something like libnoise inside a untrusted web proxy, possibly provided by the same server that hosts the web app.

I mean, instead of an untrusted web proxy, why wouldn't you use an untrusted domain name. The noise solution would be great for mobile and hopefully one day we could convince browsers to support it for Pkarr TLDs, but it is not worth it at the moment.

> I've been looking at Iroh

Iroh is using QUIC with self-signed certs, which obviously doesn't work in browser, they also tried using noise over QUIC, but it is hard work and hard to verify its security, so they are just sticking with self-signed certs and TLS for now.

Also Iroh doesn't help you if you already have a publicly accessible port.

Just point Pkarr to a domain to get the TLS support for now.

**Reply by dpc** [reply]
@Nuhvi 

> The way we do it,

I don't know. This e.g. could be a problem/weird if we ever support any other public-key based name resolution. But both schemes work just fine, so whatevs.

>  why wouldn't you use an untrusted domain name.

If we assume the DNS can be taken over, then the attacker possibly could also get a valid TLS cert for it. E2E authentication of the connection would help.

But yes - for near future seems like just being able to rotate to a different domain and using Pkarr as a globally reachable indirection system is much simpler while in practice being almost as good.
 
> Also Iroh doesn't help you if you already have a publicly accessible port.

I'd love this not being a requirement either. People could host private (behind an API secret) Federations from their desktop computers. Connectivity could work even from behind firewalls etc. It would then work like Tor hidden services, just without the Onioning part, but more reliable and performant.

**Reply by Nuhvi** [reply]
> I don't know. This e.g. could be a problem/weird if we ever support any other public-key based name resolution. But both schemes work just fine, so whatevs.

Sure. I just trust that there will never be an alternative worth using (multi-million nodes networks don't pop into existence every year), but even if one emerged and superceeded Mainline, Pkarr can publish and resolve to and from both in parallel. Publishing to mainline itself is already a parallelized operation.

> If we assume the DNS can be taken over, then the attacker possibly could also get a valid TLS cert for it. E2E authentication of the connection would help.

Fair point. I care deeply about URLs working in address bars, that's why I err on the side of normality. That being said, using normal domain can be your untrusted proxy and then you layer the e2e auth on top, when you are ready to do that extra work. And you can even make it optional so applications that is ok with TLS (like the address bar) can fallback to it.

> I'd love this not being a requirement either. People could host private (behind an API secret) Federations from their desktop computers. Connectivity could work even from behind firewalls etc. It would then work like Tor hidden services, just without the Onioning part, but more reliable and performant.

I suspect there are no human beings that are ready to take the responsibility of running a fedimint but can't open a port. They can also use dynamic DNS if they don't have static IP.

Anyway, while Iroh has holepunching, you would have to do all the wiring to connect it to web server, because they don't have that, yet. Maybe wait until they figure the wasm story, but I don't think they can give you a timeline on that.

**Reply by Nuhvi** [reply]
> I don't know. This e.g. could be a problem/weird if we ever support any other public-key based name resolution. But both schemes work just fine, so whatevs.

Another answer to this is: if I ever need a non-ed25519 key or maybe just want to migrate away from Mainline to another network exclusively, I can always make the key 53 characters instead of 52.

The logic here is: don't add distinguishing bytes until you actually have multiple choices, and keep the default simple.

**Reply by dpc** [reply]
> I suspect there are no human beings that are ready to take the responsibility of running a fedimint but can't open a port. They can also use dynamic DNS if they don't have static IP.

One day(TM) it should all be user-proof, and maybe you'll get a Bitcoin-in-a-box device, that you plug into your router and stuff just works. Also - it has other benefits, even for power users.

>  Maybe wait until they figure the wasm story

Yeah. That's the plan I guess.

**Reply by Nuhvi** [reply]
When that perfection is achieved, then it is easy to add Iroh node as a record, I mean, that is how Iroh already does it https://iroh.computer/blog/iroh-global-node-discovery#trying-it-out
